imagedephi/redact.py

```diff
@@ -1,74 +1,166 @@
-            if not tag.isIFD():
-                if tag.value in tags_to_redact:
-                    redact_one_tag(ifd, tag, tags_to_redact[tag.value])
-            else:
-                # tag_info['ifds'] contains a list of lists
-                # see tifftools.read_tiff
-                for sub_ifds in tag_info.get("ifds", []):
-                    redact_tiff_tags(sub_ifds, tags_to_redact)
-
-
-def redact_one_image(tiff_info: TiffInfo, output_path: Path) -> None:
-    ifds = tiff_info["ifds"]
-    tags_to_redact = get_tags_to_redact()
-    redact_tiff_tags(ifds, tags_to_redact)
-    tifftools.write_tiff(tiff_info, output_path)
+class TiffMetadataRedactionPlan:
+    """
+    Represents a plan of action for redacting metadata from TIFF images.
+
+    The plan can be used for reporting to end users what steps will be made to redact their TIFF
+    images, and also executing the plan.
+    """
+
+    redaction_steps: dict[int, tuple[TiffMetadataRule, RuleSource]]
+    no_match_tags: list[tifftools.TiffTag]
+    image_data: TiffInfo
+    base_rules: list[TiffMetadataRule]
+    override_rules: list[TiffMetadataRule]
+
+    @staticmethod
+    def _iter_tiff_tag_entries(
+        ifds: list[IFD],
+    ) -> Generator[tuple[tifftools.TiffTag, IFD], None, None]:
+        for ifd in ifds:
+            for tag_id, entry in sorted(ifd["tags"].items()):
+                tag: tifftools.TiffTag = tifftools.constants.get_or_create_tag(
+                    tag_id, datatype=tifftools.Datatype[entry["datatype"]]
+                )
+                if not tag.isIFD():
+                    yield tag, ifd
+                else:
+                    # entry['ifds'] contains a list of lists
+                    # see tifftools.read_tiff
+                    for sub_ifds in entry.get("ifds", []):
+                        yield from TiffMetadataRedactionPlan._iter_tiff_tag_entries(sub_ifds)
+
+    def _add_tag_to_plan(self, tag: tifftools.TiffTag) -> None:
+        """Determine how to handle a given tag."""
+        for rule in self.override_rules:
+            if rule.is_match(tag):
+                self.redaction_steps[tag.value] = (rule, RuleSource.OVERRIDE)
+                return
+        for rule in self.base_rules:
+            if rule.is_match(tag):
+                self.redaction_steps[tag.value] = (rule, RuleSource.BASE)
+                return
+        self.no_match_tags.append(tag)
+
+    def _build_redaction_steps(self, ifds: list[IFD]) -> None:
+        for tag, _ in self._iter_tiff_tag_entries(ifds):
+            self._add_tag_to_plan(tag)
+
+    def __init__(
+        self, tiff_info: TiffInfo, base_rules: RuleSet, override_rules: RuleSet | None
+    ) -> None:
+        self.redaction_steps = {}
+        self.no_match_tags = []
+        self.image_data = tiff_info
+        self.base_rules = [
+            rule for rule in base_rules.rules[RuleFormat.TIFF] if isinstance(rule, TiffMetadataRule)
+        ]
+        if override_rules is None:
```

Comment:
Is there value in this class "knowing" that `override_rules` is non-existent? If not, just make `override_rules` be required as a list, and make it the responsibility of the caller to pass an empty list. That will simplify a lot of this logic.

Reply:
See 19899e1 for changes here.

I've changed how the `TiffMetadataRedactionPlan` constructor works to simplify it, so the relevant logic is now elsewhere. The caller now passes a list of `TiffMetadataRules`, so your suggestion has been implemented.